pro/src/api/hooks/useAdapter.ts

```diff
@@ -0,0 +1,58 @@
+import React, { useEffect, useState } from 'react'
+
+type TUseAdapterLoading = [
+  data: any,
+  isLoading: true,
+  isOk: false,
+  error: undefined
+]
+
+type TUseAdapterSuccess<T> = [
+  data: T,
+  isLoading: false,
+  isOk: true,
+  error: undefined
+]
+
+type TUseAdapterFailure<T> = [
+  data: T,
+  isLoading: false,
+  isOk: false,
+  error: string
+]
+
+const useAdapter = <ISuccessPayload, IFailurePayload>(
+  getData: () => Promise<
+    AdapterSuccess<ISuccessPayload> | AdapterFailure<IFailurePayload>
+  >,
+  defaultValue: any
+):
+  | TUseAdapterLoading
+  | TUseAdapterSuccess<ISuccessPayload>
+  | TUseAdapterFailure<IFailurePayload> => {
+  const [hookResponse, setHookResponse] = useState<
+    | TUseAdapterLoading
+    | TUseAdapterSuccess<ISuccessPayload>
+    | TUseAdapterFailure<IFailurePayload>
+  >([defaultValue, true, false, undefined])
```

Comment:
```suggestion
  const [hookResponse, setHookResponse] = useState<
    ReturnType<typeof useAdapter>
  >([defaultValue, true, false, undefined])
```

Comme ça c'est plus simple, on rend évident le fait que c'est le même type.
Et c'est une bonne pratique Typescript de réutiliser les types des variables : par exemple si plus tard on change le return type de la fonction, ton hookResponse aura toujours le bon type, et de possible erreurs seront détectées plus bas

Reply:
Pour le point sur `IFailurePayload` qui est inutile alors que `ISuccessPayload` l'est, j'ai du mal à saisir.

Reply:
Hum doublement en effet,
pour le montage de ma suggestion ça ne marche pas, j'ai trouvé quelques solutions mais ça commence à être du bricolage oublions

Reply:
pour le fun j'aime bien le bricolage, même si ça n'est pas mergabe :)

Reply:
Oui moi aussi :)

Du coup comme solution il y a : 

1) ton `ReturnType<typeof useAdapter<ISuccessPayload, IFailurePayload>>`, je pense que c'est le mieux en fait, d'après cette discussion SO (https://github.com/microsoft/TypeScript/pull/47607) c'est dispo sur la version mineure de TS juste au dessus de la notre,
à reprendre quand on upgrade je pense

2) sur cette même issue SO ils proposent de définir le type de la fonction par une interface, qui permet de ne pas utiliser typeof pour faire la même opération qu'à la 1 et donc ça marche déjà, mais vu le code c'est trop bricolage je pense : 

```
interface useAdapter<ISuccessPayload, IFailurePayload> {
  (
    getData: () => Promise<
      AdapterSuccess<ISuccessPayload> | AdapterFailure<IFailurePayload>
    >
  ):
    | TUseAdapterLoading
    | TUseAdapterSuccess<ISuccessPayload>
    | TUseAdapterFailure<IFailurePayload>
}

const useAdapter = <ISuccessPayload, IFailurePayload>(
  ...[getData]: Parameters<useAdapter<ISuccessPayload, IFailurePayload>>
): ReturnType<useAdapter<ISuccessPayload, IFailurePayload>> => {

  const [hookResponse, setHookResponse] = useState<
    ReturnType<useAdapter<ISuccessPayload, IFailurePayload>>
  >([undefined, true, false, undefined])
```

3) ça revient un peu à définir le type de retour avant et de l'utiliser dans la fonction, c'est le plus clean avec le TS actuel je pense : ne pas passer par les TUseAdapterXX mais juste définir 
```
type TUseAdapterReturn = [
  data: undefined,
  isLoading: true,
  isOk: false,
  error: undefined
] | [
  data: T,
  isLoading: false,
  isOk: true,
  error: undefined
] | [
  data: T,
  isLoading: false,
  isOk: false,
  error: string
]
```
ça revient quand même au même que la version actuelle, avec plus ou moins de répétition suivant le goût de chacun